### PR TITLE
feat: Converted project to typescript

### PR DESCRIPTION
**Pull request by wbreza**
- Converts the project over to typescript
- Created new `Login` plugin to handle azure login
- Create new `APIM` plugin for integration with Azure API management
- Started effort of refactoring `Deploy` plugin

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,10 +3,12 @@
   "version": "0.7.0",
   "description": "Provider plugin for the Serverless Framework v1.x which adds support for Azure Functions.",
   "license": "MIT",
-  "main": "index.js",
+  "main": "./lib/index.js",
   "author": "Azure Functions",
   "scripts": {
-    "test": "eslint **/*.js"
+    "test": "eslint src/**/*.js",
+    "prebuild": "rm lib/ -rf && npm test",
```
Comment (mydiemho) [inline package.json]: nit:`rm -rf lib/`

File: src/config.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
-module.exports = {
+export const constants = {
   bearer: 'Bearer ',
   functionAppApiPath: '/api/',
   functionAppDomain: '.azurewebsites.net',
-  functionsAdminApiPath: '.azurewebsites.net/admin/functions/',
+  functionsAdminApiPath: '/admin/functions/',
```
Comment (mydiemho) [inline src/config.ts]: are function and functions referring to 2 different things?

File: src/config.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
   functionsApiPath: '/api/functions',
   jsonContentType: 'application/json',
   logInvocationsApiPath: '/azurejobs/api/functions/definitions/',
```
Comment (mydiemho) [inline src/config.ts]: consistency: some path end with `/` and some don't

File: src/config.ts [hunk 4]
```diff
@@ -12,6 +12,8 @@ module.exports = {
   providerName: 'azure',
   scmCommandApiPath: '/api/command',
   scmDomain: '.scm.azurewebsites.net',
-  scmVfsPath: '.scm.azurewebsites.net/api/vfs/site/wwwroot/',
-  scmZipApiPath: '.scm.azurewebsites.net/api/zip/site/wwwroot/'
+  scmVfsPath: '/api/vfs/site/wwwroot/',
+  scmZipApiPath: '/api/zip/site/wwwroot/'
 };
+
+export default constants;
```
Comment (mydiemho) [inline src/config.ts]: style: new line in file or not? we should decide on one and add to editorconfig

File: src/plugins/deploy/azureDeployPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import { ResourceService } from '../../services/resourceService';
+import { FunctionAppService } from '../../services/functionAppService';
+
+export class AzureDeployPlugin {
+  hooks: any;
```
Comment (mydiemho) [inline src/plugins/deploy/azureDeployPlugin.ts]: private?

File: src/plugins/invoke/azureInvoke.ts [hunk 6]
```diff
@@ -1,28 +1,30 @@
...
-class AzureInvoke {
-  constructor (serverless, options) {
+export class AzureInvoke {
+  hooks: any;
+  provider: any;
+  invokeFunction: any;
+  getAdminKey: any;
+
+  constructor (private serverless, private options) {
     this.serverless = serverless;
```
Comment (mydiemho) [inline src/plugins/invoke/azureInvoke.ts]: can remove

File: src/plugins/invoke/azureInvoke.ts [hunk 7]
```diff
@@ -1,28 +1,30 @@
-'use strict';
 
-const BbPromise = require('bluebird');
+import { Promise } from 'bluebird';
```
Comment (mydiemho) [inline src/plugins/invoke/azureInvoke.ts]: don't need bluebird anymore

File: src/plugins/logs/azureLogs.ts [hunk 8]
```diff
@@ -1,31 +1,30 @@
-'use strict';
 
-const BbPromise = require('bluebird');
+import { Promise } from 'bluebird';
```
Comment (mydiemho) [inline src/plugins/logs/azureLogs.ts]: don't need bluebird

File: src/plugins/package/azurePackage.ts [hunk 9]
```diff
@@ -1,11 +1,16 @@
-'use strict';
 
-const BbPromise = require('bluebird');
+import { Promise } from 'bluebird';
```
Comment (mydiemho) [inline src/plugins/package/azurePackage.ts]: don't need bluebird

File: src/plugins/package/azurePackageFunction.ts [hunk 10]
```diff
@@ -1,10 +1,14 @@
-'use strict';
 
-const BbPromise = require('bluebird');
+import { Promise } from 'bluebird';
```
Comment (mydiemho) [inline src/plugins/package/azurePackageFunction.ts]: don't need bluebird

File: src/plugins/package/lib/compileEvents.ts [hunk 11]
```diff
@@ -1,7 +1,6 @@
-'use strict';
 
-const BbPromise = require('bluebird');
-const utils = require('../../shared/utils');
+import { Promise } from 'bluebird';
```
Comment (mydiemho) [inline src/plugins/package/lib/compileEvents.ts]: don't need bluebird

File: src/plugins/package/lib/webpackFunctionJson.ts [hunk 12]
```diff
@@ -1,8 +1,6 @@
-'use strict';
 const fs = require('fs');
-const path = require('path');
-const BbPromise = require('bluebird');
-
+import { join } from 'path';
+import { Promise } from 'bluebird';
```
Comment (mydiemho) [inline src/plugins/package/lib/webpackFunctionJson.ts]: don't need bluebird

File: package.json (hunk 1 above)
Comment (wbreza) [reply package.json]: Any advantage of putting flags before?  My habit is to put them after.

File: src/config.ts (hunk 2 above)
Comment (wbreza) [reply src/config.ts]: This is a partial path to one of the Kudu rest endpoints to retrieve functions in a function app.

File: src/config.ts (hunk 3 above)
Comment (tbarlow12) [reply src/config.ts]: I think it's because some of these will be appended to others. For example `{functionAppDomain}` + `{functionsApiPath}`

**Pull request merged**